### ISSUE TITLE
CodeCoverage is always enabled when configuring it with a .runsettings file, independend of /EnableCodeCoverage

### ISSUE THREAD

**Issue by Mertsch** [Microsoft/vstest#1278]
## Description
I have a project with a `.runsettings` to configure code coverage

```
<DataCollectionRunSettings>
  <DataCollectors>
    <!-- https://docs.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis -->
    <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0"
                   assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
      <Configuration>
        <CodeCoverage>
          <ModulePaths>
            <Exclude>
              <ModulePath>.*FakeItEasy.*</ModulePath>
              <ModulePath>.*GongSolutions.WPF.DragDrop.*</ModulePath>
            </Exclude>
          </ModulePaths>
        </CodeCoverage>
      </Configuration>
    </DataCollector>
  </DataCollectors>
</DataCollectionRunSettings>
```

In VSTS I have set up a Gated build and a CI build (which runs after Gated)
Gated will run only the impacted tests, CI all

I noticed that even though I have set my Gated build **not** to run Code Coverage, I still get CC at the end of the Build.

Looking through the code I figured that each collector has a flag `IsEnabled` https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectorSettings.cs#L46
and if any is, that data collection will happen https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/DataCollector/DataCollectionRunSettings.cs#L71

This leads to `vstest.console.exe` ignoring the `/EnableCodeCoverage` parameter.
I tried disabling the `DataCollector` via `enabled=false`, which leads to no code coverage independent of `/EnableCodeCoverage`

## Steps to reproduce
See description

## Expected behavior
`/EnableCodeCoverage` should enable/disable the CC data collector or there should be some means of configuring the data collector without automatically enabling it

## Actual behavior
Data collection can not be configured without specifically enabling/disabling it.

## Environment
Visual Studio 2017 15.4.2
.NET Framework 4.6.1 project

**Comment by harshjain2** [comment]
@Mertsch you can use `enabled="false"` attribute in .runsettings to disable a CodeCoverage DataCollector by default. You can enable it by using `/EnableCodeCoverage` or `/collect:"Code Coverage" command line arguments.

More info [here](https://github.com/Microsoft/vstest-docs/blob/master/docs/analyze.md#enabledisable-a-datacollector).

**Comment by harshjain2** [comment]
I used the below runsettings
```xml
<RunSettings>  
  <DataCollectionRunSettings>
  <DataCollectors>
    <!-- https://docs.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis -->
    <DataCollector friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0"
                   assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" enabled="false">
      <Configuration>
        <CodeCoverage>
          <ModulePaths>
            <Exclude>
              <ModulePath>.*FakeItEasy.*</ModulePath>
              <ModulePath>.*GongSolutions.WPF.DragDrop.*</ModulePath>
            </Exclude>
          </ModulePaths>
        </CodeCoverage>
      </Configuration>
    </DataCollector>
  </DataCollectors>
</DataCollectionRunSettings>
</RunSettings>
```
And this command, I don't get any CodeCoverage report.
> C:\code\github\harshjain2\vstest>"C:\code\github\harshjain2\vstest\artifacts\Debug\packages\Microsoft.TestPlatform.15.6.0-dev\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe" .\test\TestAssets\SimpleTestProject2\bin\Debug\net451\SimpleTestProject2.dll /settings:C:\Users\harjain\Desktop\rs.runsettings


But when I use this command I get CodeCoverage report.
>C:\code\github\harshjain2\vstest>"C:\code\github\harshjain2\vstest\artifacts\Debug\packages\Microsoft.TestPlatform.15.6.0-dev\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe" .\test\TestAssets\SimpleTestProject2\bin\Debug\net451\SimpleTestProject2.dll /settings:C:\Users\harjain\Desktop\rs.runsettings /EnableCodeCoverage

**Comment by Mertsch** [comment]
@harshjain2 I did the same thing and there is no code coverage

```
2017-11-11T16:35:50.5001676Z ##[section]Starting: Test Assemblies
2017-11-11T16:35:50.5001676Z ==============================================================================
2017-11-11T16:35:50.5001676Z Task         : Visual Studio Test
2017-11-11T16:35:50.5001676Z Description  : Run tests with Visual Studio test runner
2017-11-11T16:35:50.5001676Z Version      : 2.2.3
2017-11-11T16:35:50.5001676Z Author       : Microsoft Corporation
2017-11-11T16:35:50.5001676Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=835764)
2017-11-11T16:35:50.5001676Z ==============================================================================
2017-11-11T16:35:52.6064502Z Run the tests locally using vstest.console.exe
2017-11-11T16:35:52.6064502Z ========================================================
2017-11-11T16:35:52.6064502Z Test selector : Test assemblies
2017-11-11T16:35:52.6064502Z Test assemblies : **\*tests.dll,!**\obj\**,!**\packages\**
2017-11-11T16:35:52.6064502Z Test filter criteria : null
2017-11-11T16:35:52.6064502Z Search folder : C:\Build\_work\7\s
2017-11-11T16:35:52.6064502Z Run settings file : C:\Build\_work\7\s\Default.runsettings
2017-11-11T16:35:52.6064502Z Run in parallel : false
2017-11-11T16:35:52.6064502Z Run in isolation : true
2017-11-11T16:35:52.6064502Z Path to custom adapters : null
2017-11-11T16:35:52.6064502Z Other console options : null
2017-11-11T16:35:52.6064502Z Code coverage enabled : true
2017-11-11T16:35:52.6064502Z VisualStudio version selected for test execution : 15.0
2017-11-11T16:35:53.6242932Z ========================================================
2017-11-11T16:35:56.3898627Z [command]"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" @C:\Build\_work\_temp\624a5181-c6fe-11e7-8b4c-61552c82644e.txt
2017-11-11T16:35:56.6623766Z Microsoft (R) Test Execution Command Line Tool Version 15.0.26929.2
2017-11-11T16:35:56.6623766Z Copyright (c) Microsoft Corporation.  All rights reserved.
2017-11-11T16:35:56.6623766Z 
2017-11-11T16:35:56.6623766Z vstest.console.exe 
2017-11-11T16:35:56.6623766Z "C:\Build\_work\7\s\Project.PDF.Tests\bin\Release\Project.PDF.Tests.dll"
2017-11-11T16:35:56.6623766Z "C:\Build\_work\7\s\Project.Tests\bin\Release\Project.Tests.dll"
2017-11-11T16:35:56.6623766Z /Settings:"C:\Build\_work\7\s\Default.runsettings"
2017-11-11T16:35:56.6623766Z /EnableCodeCoverage
2017-11-11T16:35:56.6623766Z /InIsolation
2017-11-11T16:35:56.6623766Z /logger:"trx"
2017-11-11T16:35:56.6623766Z /TestAdapterPath:"C:\Build\_work\7\s"
2017-11-11T16:35:57.2429794Z Starting test execution, please wait...
2017-11-11T16:36:03.3779983Z Passed   [...]
2017-11-11T16:36:40.8104176Z Results File: C:\Build\_work\7\s\TestResults\VSTSBuildAgent_BUILDSERVER 2017-11-11 16_36_03.trx
2017-11-11T16:36:40.8104176Z 
2017-11-11T16:36:40.8260419Z Total tests: 30. Passed: 30. Failed: 0. Skipped: 0.
2017-11-11T16:36:40.8260419Z Test Run Successful.
2017-11-11T16:36:40.8260419Z Test execution time: 43.3347 Seconds
2017-11-11T16:36:41.5209118Z ##[section]Async Command Start: Publish test results
2017-11-11T16:36:41.6416167Z Publishing test results to test run '7386'
2017-11-11T16:36:41.6416167Z Test results remaining: 30. Test run id: 7386
2017-11-11T16:36:44.0154675Z Published Test Run : https://account.visualstudio.com/Project/_TestManagement/Runs#runId=7386&_a=runCharts
2017-11-11T16:36:44.0154675Z ##[section]Async Command End: Publish test results
2017-11-11T16:36:44.0154675Z ##[section]Finishing: Test Assemblies
```

using
```
<package id="MSTest.TestAdapter" version="1.2.0" targetFramework="net461" />
<package id="MSTest.TestFramework" version="1.2.0" targetFramework="net461" />
```

**Comment by mayankbansal018** [comment]
@Mertsch I suppose the issue is resolved for you, can you please confirm.

**Comment by Mertsch** [comment]
@mayankbansal018 Did I miss something? There was no update and `enabled="false"` + `/EnableCodeCoverage` still leads to no code coverage and is no solution.
If there is no CC in `.runsettings` everything works fine.

**Comment by mayankbansal018** [comment]
@Mertsch behind the scenes /EnableCodeCoverage flag simply tries to add some default CodeCoverage configuration to runsettings if there is nothing in runsettings which would enable CodeCoverage.

In your case the runsettings file already contains the CodeCoverage section, so /EnableCodeCoverage flag will do nothing, it honors whatever setting user has mentioned in runsettings file, which in this case says that do not do code coverage as you have marked enabled = false.

I think the name /EnableCodeCoverage is misleading here causing users to assume the if it is set then codecoverage will always happen, which is not true in this specific case.

So for this case the solution would be that you only provide runsettings which contain code coverage settings only when you want it.

**Comment by Mertsch** [comment]
@mayankbansal018 I understand, would be nice to see this behavior changed.
I guess it would be a simple change to detect an existing CC setting and set the `IsEnabled` flag to true.
Having to maintain two different `.runsettings` files with (almost) identical content seems a rather bad solution.
Also your colleague said in https://github.com/Microsoft/vstest/issues/1278#issuecomment-343663679 that it works for him. I see he is using some kind of preview version. Maybe it will be resolved in VS 2017 15.5 / 15.6 ?

**Comment by mayankbansal018** [comment]
@Mertsch I just tried out what mycolleague said in [#1278 (comment)](https://github.com/Microsoft/vstest/issues/1278#issuecomment-343663679) , & it works fine in VS 2017.5, the issue should get resolved for you once you move to latest VS 2017.5

I'll close this issue, please do reopen if you still face the problem after upgrading to latest VS

**Comment by Mertsch** [comment]
@mayankbansal018 Thank you for the clarification, I will Upgrade to 15.5 as soon as it's released and try again.
